benchmarks/htap/lib/stats.py

```diff
@@ -68,18 +96,34 @@ def __init__(self, dsn, num_oltp_slots, num_olap_slots, csv_interval, ignored_qu
         self.csv_oltp = None
         self.csv_dbstats = None
         self.conn = None
+        self.history_length = history_length
+
+    def get_history_length(self):
+        return len(self.data['oltp']['history'])
 
     def _update_oltp_stats(self, stats):
         for stat in stats:
-            self.data['oltp'][stat['query']].append(stat)
             self.data['oltp_counts'][stat['query']][stat['status']] += 1
 
-
-    def cleanup_oltp_stats(self, time_now):
-        cutoff = int(time_now - HISTORY_LENGTH)
-        for query_type in self.data['oltp'].keys():
-            # keep HISTORY_LENGTH of history
-            self.data['oltp'][query_type] = [s for s in self.data['oltp'][query_type] if s['timestamp'] >= cutoff]
+            oltp = self.data['oltp']
+            second = int(stat['timestamp'])
+            if second < oltp['history'][0][0]:
+                # Too old sample, discard it
+                continue
+            if second > oltp['history'][-1][0]:
+                # Sample goes beyond the horizon, expand the bucket list to accomodate all these seconds
+                base = oltp['history'][-1][0]
+                oltp['history'].extend((base + i, {k:OLTPBucketStats() for k in QUERY_TYPES} ) for i in range(1, 1 + second - base))
+
+                # Remove older buckets if we go beyond the maximum history size
+                excess = len(oltp['history']) - oltp['max_history']
```

Comment:
so what is the difference between history_length and max_history? maybe just use one of the vars?

Reply:
changed the underlying struct to a deque, so I do not need to keep the history_length around, and I do not have to manually drop values on the left of the history when it is extended